src/app/sidebar/sidebar/sidebar.component.ts

```diff
@@ -114,38 +120,159 @@ export class SidebarComponent {
+      mapStyle: undefined,
+    };
+
+    await db.localMapMeta.put(localMapMeta);
+
     let localMapBlob = await db.localMapBlobs.get(localMapMeta.url);
     if (!localMapBlob) {
-      try {
-        const [localMap, mapStyle] = await Promise.all([
-          lastValueFrom(this.http.get(downloadUrl, { responseType: 'blob' })),
-          fetch('/assets/map-style.json').then((res) => res.text()),
-        ]);
-        localMapBlob = {
-          url: localMapMeta.url,
-          data: localMap,
-        };
-        await db.localMapBlobs.add(localMapBlob);
-        localMapMeta.mapStyle = mapStyle;
-        localMapMeta.objectUrl = undefined;
-        localMapMeta.mapStatus = 'downloaded';
-      } catch (e) {
+      const mapRequest = this.http
+        .get(downloadUrl, {
+          responseType: 'blob',
+          reportProgress: true,
+          observe: 'events',
+        })
+        .subscribe({
+          next: async (event) => {
+            if (event.type === HttpEventType.DownloadProgress) {
+              const percentDone = Math.round((100 * event.loaded) / (event.total ?? 1));
+              this.mapProgress = percentDone;
+              this.cdRef.detectChanges();
+            } else if (event instanceof HttpResponse) {
+              const localMap = event.body as Blob;
+              const mapStyle = await fetch('/assets/map-style.json').then((res) => res.text());
+              localMapBlob = {
+                url: localMapMeta.url,
+                data: localMap,
+              };
+              await db.localMapBlobs.add(localMapBlob);
+              localMapMeta.mapStyle = mapStyle;
+              localMapMeta.objectUrl = undefined;
+              localMapMeta.mapStatus = 'downloaded';
+              this.mapDownloadStates[map] = localMapMeta.mapStatus;
+              await db.localMapMeta.put(localMapMeta);
+              this.cdRef.detectChanges();
+            }
+          },
+          error: async () => {
+            localMapMeta.mapStatus = 'missing';
+            this.mapProgress = 0;
+            this.mapDownloadStates[map] = localMapMeta.mapStatus;
+            await db.localMapMeta.put(localMapMeta);
+          },
+        });
+
+      // Store the subscription so it can be cancelled later
+      this.subscriptions.set(map, mapRequest);
+    }
+  }
+
+  async cancelDownloadMap(map: ZsMapStateSource) {
+    this.mapDownloadStates[map] = 'missing';
+    if (this.subscriptions.has(map)) {
+      this.subscriptions.get(map)?.unsubscribe();
+      this.subscriptions.delete(map);
+      this.mapProgress = 0;
+      const downloadUrl = zsMapStateSourceToDownloadUrl[map];
+      const localMapMeta = await db.localMapMeta.get(downloadUrl);
+      if (localMapMeta) {
         localMapMeta.mapStatus = 'missing';
+        await db.localMapMeta.put(localMapMeta);
       }
     }
-    this.mapDownloadStates[map] = localMapMeta.mapStatus;
+  }
+
+  async uploadMap(event: Event, map: ZsMapStateSource) {
```

Comment:
quite a lot of code duplication in this function. maybe some parts can be extracted into a dedicated function